transport/src/test/java/io/netty/channel/SingleThreadEventLoopTest.java

```diff
@@ -499,4 +558,109 @@ protected void wakeup(boolean inEventLoop) {
+
+        final LinkedBlockingQueue<Boolean> iterationEndSignal = new LinkedBlockingQueue<Boolean>(1);
+        final AtomicInteger runtTimedOutCount = new AtomicInteger();
+        private final LinkedBlockingQueue<Boolean> runSignal = new LinkedBlockingQueue<Boolean>(1);
+        private final int timeoutNanos;
+        private volatile boolean ignoreWakeup;
+        private volatile boolean shutdown;
+
+        SingleThreadEventLoopC(int timeoutNanos) {
+            super(null, Executors.defaultThreadFactory(), false);
+            this.timeoutNanos = timeoutNanos;
+        }
+
+        @Override
+        protected void run() {
+            while (!shutdown) {
+                try {
+                    runSignal.take();
+                } catch (InterruptedException e) {
+                    // Waken up by interruptThread()
+                }
+
+                runAllTasks(timeoutNanos);
+                iterationEndSignal.offer(true);
+            }
+
+            confirmShutdown();
+        }
+
+        @Override
+        protected void wakeup(boolean inEventLoop) {
+            if (!ignoreWakeup) {
+                runNextIteration();
+            }
+        }
+
+        public void runNextIteration() {
+            runSignal.offer(true);
+        }
+
+        public void signalShutdown() {
+            shutdown = true;
+            setIgnoreWakeup(false);
+            wakeup(false);
+        }
+
+        public void setIgnoreWakeup(boolean ignore) {
+            ignoreWakeup = ignore;
+        }
+
+        @Override
+        protected void afterRunningAllTasks(boolean timedOut) {
+            if (timedOut) {
+                runtTimedOutCount.incrementAndGet();
+            }
+            super.afterRunningAllTasks(timedOut);
+        }
+    }
+
+    private static class CountingRunnable implements Runnable {
+
+        private final AtomicInteger invocationCount = new AtomicInteger();
+
+        @Override
+        public void run() {
+            invocationCount.incrementAndGet();
+        }
+
+        public int getInvocationCount() {
+            return invocationCount.get();
+        }
+
+        public void resetInvocationCount() {
+            invocationCount.set(0);
+        }
+    }
+
+    private static final class DelayTask extends CountingRunnable {
+
+        private final LinkedBlockingQueue<Boolean> resumeSignal = new LinkedBlockingQueue<Boolean>(1);
+        private final LinkedBlockingQueue<Boolean> runningSignal = new LinkedBlockingQueue<Boolean>(1);
+
+        @Override
+        public void run() {
+            runningSignal.offer(true);
+            try {
+                resumeSignal.take();
+            } catch (InterruptedException e) {
+                // Waken up by interruptThread()
+            }
+            super.run();
+        }
+
+        public void resumeAfter(int awaitNanos) {
+            try {
+                runningSignal.take();
+                Thread.sleep(0, awaitNanos);
```

Comment:
This is required because I am introducing delay to cause timeout of `runAllTasks()`